File: core/src/main/java/com/jlpka/langidentify/WordSegmenter.java
/*
 * Copyright 2026 Jeremy Lilley (jeremy@jlilley.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.jlpka.langidentify;

import java.io.IOException;
import java.io.Reader;

/**
 * Segments text into words by alphabet, handling lowercasing, apostrophe splitting, digit
 * filtering, and CJ run detection. Calls back to consumers for each emitted word and CJ run.
 */
class WordSegmenter {
  /** Maximum length for alphabetic words; longer words are truncated. */
  public static final int MAX_WORD_LEN = 64;

  /** Maximum length for CJ runs; longer runs are split into chunks of this size. */
  public static final int MAX_CJ_WORD_LEN = 256;

  static final int HAS_DIGIT = 1;
  static final int HAS_APOSTROPHE = 2;

  @FunctionalInterface
  interface WordConsumer {
    void accept(char[] wordBuf, int wordLen, int alphaIdx);
  }

  private final Model model;
  final char[] wordBuf = new char[MAX_CJ_WORD_LEN];
  private final WordConsumer wordConsumer;
  private final WordConsumer cjConsumer; // null if no CJ classification needed

  // Mutable state, reset on each segment() call
  private int wordLen;
  private int wordStart; // start index in original text, -1 if no current word
  private int wordAlphaIdx;
  private int specialCases;

  WordSegmenter(Model model, WordConsumer wordConsumer, WordConsumer cjConsumer) {
    this.model = model;
    this.wordConsumer = wordConsumer;
    this.cjConsumer = cjConsumer;
  }

  void segment(CharSequence text) {
    initSegment();
    final int length = text.length();
    for (int i = 0; i < length; i++) {
      handleChar(text.charAt(i), i);
    }
    if (wordStart >= 0) {
      emitWord();
    }
  }

  void segment(char[] text, int ofs, int len) {
    initSegment();
    for (int i = 0; i < len; i++) {
      handleChar(text[ofs + i], i);
    }
    if (wordStart >= 0) {
      emitWord();
    }
  }

  void segment(Reader reader) throws IOException {
    initSegment();
    int pos = 0;
    char[] buf = new char[1024];
    for (; ; ) {
      int nread = reader.read(buf);
      if (nread <= 0) break;
      for (int i = 0; i < nread; i++) {
        handleChar(buf[i], pos++);
      }
    }
    if (wordStart >= 0) {
      emitWord();
    }
  }

  private void initSegment() {
    wordLen = 0;
    wordStart = -1;
    wordAlphaIdx = -1;
    specialCases = 0;
  }

  private void handleChar(char c, int pos) {
    Alphabet alpha = Alphabet.getAlphabet(c);
    int alphaIdx = model.alphabetIndex(alpha);

    if (alphaIdx >= 0) {
      if (alphaIdx != wordAlphaIdx && wordStart >= 0) {
        emitWord();
      }
      if (wordLen < MAX_WORD_LEN) {
        wordBuf[wordLen++] = Character.toLowerCase(c);
      } else { // uncommon
        if (model.isCJAlphabet(
            wordAlphaIdx)) { // CJ special case: higher limit + emit the partial word.
          if (wordLen == MAX_CJ_WORD_LEN) {
            emitWord();
          }
          wordBuf[wordLen++] = c;
        }
      }
      if (wordStart < 0) wordStart = pos;
      wordAlphaIdx = alphaIdx;
    } else if (c >= '0' && c <= '9') { // only care about standard digits here
      specialCases |= HAS_DIGIT;
    } else if (wordStart >= 0
        && (c == '\'' || c == '\u2019' || c == '\u0092')
        && wordBuf[wordLen - 1] != '\'') {
      // \u2019 = right single quotation mark (curly apostrophe).
      // \u0092 = Windows-1252 apostrophe mis-decoded as UTF-8.
      specialCases |= HAS_APOSTROPHE;
      if (wordLen < MAX_WORD_LEN) {
        wordBuf[wordLen++] = '\'';
      }
    } else {
      if (wordStart >= 0) {
        emitWord();
      }
      specialCases = 0;
    }
  }

  // pos = current position in text (boundary char, or text.length() for trailing word)
  private void emitWord() {
    // Special-case filtering
    if (specialCases != 0) {
      if ((specialCases & HAS_DIGIT) != 0 && model.getAlphabets()[wordAlphaIdx] == Alphabet.LATIN) {
        wordLen = 0;
        wordStart = -1;
        specialCases = 0;
        return;
      }
      if ((specialCases & HAS_APOSTROPHE) != 0 && wordBuf[wordLen - 1] == '\'') {
        if (--wordLen == 0) {
          wordStart = -1;
          specialCases = 0;
          return;
        }
      }
    }

    if (cjConsumer != null && model.isCJAlphabet(wordAlphaIdx)) {
      cjConsumer.accept(wordBuf, wordLen, wordAlphaIdx);
    } else {
      wordConsumer.accept(wordBuf, wordLen, wordAlphaIdx);
    }

    wordLen = 0;
    wordStart = -1;
    specialCases = 0;
  }
}

File: python/core/src/langidentify/word_segmenter.py
# Copyright 2026 Jeremy Lilley (jeremy@jlilley.net)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Segments text into words by alphabet, handling lowercasing, apostrophe
splitting, digit filtering, and CJ run detection."""

from langidentify.alphabet import Alphabet, get_alphabet

MAX_WORD_LEN = 64
MAX_CJ_WORD_LEN = 256

_HAS_DIGIT = 1
_HAS_APOSTROPHE = 2

# Apostrophe characters: ASCII ', right single quotation mark, Windows-1252 mis-decode
_APOSTROPHES = frozenset(["'", "\u2019", "\u0092"])


class WordSegmenter:
    """Segments text into words by alphabet boundaries.

    Calls word_consumer(word, alpha_idx) for each emitted word, and
    cj_consumer(word, alpha_idx) for CJ runs.
    """

    def __init__(self, model, word_consumer, cj_consumer=None):
        self._model = model
        self._word_consumer = word_consumer
        self._cj_consumer = cj_consumer

    def segment(self, text):
        """Segments a string into words."""
        word_buf = []
        word_start = -1
        word_alpha_idx = -1
        special_cases = 0
        model = self._model

        for i, ch in enumerate(text):
            alpha = get_alphabet(ch)
            alpha_idx = model.alphabet_index(alpha)

            if alpha_idx >= 0:
                if alpha_idx != word_alpha_idx and word_start >= 0:
                    self._emit_word(
                        word_buf, word_alpha_idx, special_cases)
                    word_buf = []
                    special_cases = 0

                if len(word_buf) < MAX_WORD_LEN:
                    word_buf.append(ch.lower())
                elif model.is_cj_alphabet(word_alpha_idx):
                    if len(word_buf) == MAX_CJ_WORD_LEN:
                        self._emit_word(
                            word_buf, word_alpha_idx, special_cases)
                        word_buf = []
                        special_cases = 0
                    word_buf.append(ch)

                if word_start < 0:
                    word_start = i
                word_alpha_idx = alpha_idx

            elif "0" <= ch <= "9":
                special_cases |= _HAS_DIGIT

            elif (word_start >= 0
                  and ch in _APOSTROPHES
                  and word_buf
                  and word_buf[-1] != "'"):
                special_cases |= _HAS_APOSTROPHE
                if len(word_buf) < MAX_WORD_LEN:
                    word_buf.append("'")

            else:
                if word_start >= 0:
                    self._emit_word(
                        word_buf, word_alpha_idx, special_cases)
                    word_buf = []
                special_cases = 0
                word_start = -1
                word_alpha_idx = -1

        if word_start >= 0:
            self._emit_word(word_buf, word_alpha_idx, special_cases)

    def _emit_word(self, word_buf, alpha_idx, special_cases):
        if not word_buf:
            return

        model = self._model

        # Special-case filtering
        if special_cases != 0:
            if ((special_cases & _HAS_DIGIT)
                    and model.alphabets[alpha_idx] == Alphabet.LATIN):
                return
            if (special_cases & _HAS_APOSTROPHE) and word_buf[-1] == "'":
                word_buf.pop()
                if not word_buf:
                    return

        word = "".join(word_buf)

        if self._cj_consumer is not None and model.is_cj_alphabet(alpha_idx):
            self._cj_consumer(word, alpha_idx)
        else:
            self._word_consumer(word, alpha_idx)
